File: apps/web/src/services/brightdata.ts
import crypto from 'crypto';
import { db } from '../db';
import { scrapers, selectorVersions } from '../db/schema';
import { eq, desc } from 'drizzle-orm';
import {
  fetchSandboxRecords,
  getTemplateForUrl,
  mapRecordsToSchema,
} from './sandboxScraper';

// Persist the simulator's selector state across hot reloads in development
const globalForSelectors = global as unknown as {
  simulatedSelectors: Record<string, Record<string, string>>;
};

export const SIMULATED_SELECTORS = globalForSelectors.simulatedSelectors || {};

if (process.env.NODE_ENV !== 'production') {
  globalForSelectors.simulatedSelectors = SIMULATED_SELECTORS;
}

export class BrightDataService {
  static getBaseUrl(): string {
    if (process.env.NEXT_PUBLIC_APP_URL) {
      return process.env.NEXT_PUBLIC_APP_URL;
    }
    return 'http://localhost:3000';
  }

  static isMockMode(): boolean {
    const apiKey = process.env.BRIGHTDATA_API_KEY;
    const mockEnv = process.env.MOCK_BRIGHTDATA;
    return !apiKey || mockEnv === 'true';
  }

  /** Auto-generated IDs (c_xxxxxxxx) are local-only — not real Bright Data collectors. */
  static isLocalCollector(collectorId: string): boolean {
    return !collectorId || /^c_[0-9a-f]{8}$/i.test(collectorId);
  }

  static shouldSimulate(collectorId: string): boolean {
    return this.isMockMode() || this.isLocalCollector(collectorId);
  }

  static async computePageHash(url: string): Promise<string | null> {
    try {
      let fetchUrl = url;
      if (this.isMockMode()) {
        fetchUrl = `${this.getBaseUrl()}/api/demo/catalog-raw`;
      }
      
      try {
        const res = await fetch(fetchUrl, { next: { revalidate: 0 } });
        if (res.status === 200) {
          const text = await res.text();
          return crypto.createHash('sha256').update(text).digest('hex');
        }
      } catch (fetchErr) {
        if (this.isMockMode()) {
          // Local server is offline (e.g. during cold start/build), fallback to stable hash of URL
          return crypto.createHash('sha256').update(url).digest('hex');
        }
        throw fetchErr;
      }
    } catch (e) {
      console.error('Error computing page hash:', e);
    }
    return null;
  }

  static async updateScraperCache(scraperId: string, records: any[], htmlHash: string) {
    if (!htmlHash) return;
    try {
      await db.update(scrapers)
        .set({
          lastHtmlHash: htmlHash,
          cachedRecords: records,
          updatedAt: new Date(),
        })
        .where(eq(scrapers.id, scraperId));
    } catch (e) {
      console.error('Error updating scraper cache:', e);
    }
  }

  static async runCollector(
    collectorId: string,
    targetUrl: string,
    schema: Record<string, string>,
    scraperId?: string,
    bypassCache: boolean = false
  ): Promise<{
    status: string;
    records: any[];
    error: string | null;
    cached?: boolean;
    hash?: string | null;
  }> {
    let currentHash: string | null = null;

    // 1. Pre-flight Cache Check
    if (scraperId && !bypassCache) {
      const foundList = await db.select().from(scrapers).where(eq(scrapers.id, scraperId)).limit(1);
      const scraper = foundList[0];
      if (scraper && scraper.lastHtmlHash && scraper.cachedRecords) {
        currentHash = await this.computePageHash(targetUrl);
        if (currentHash && currentHash === scraper.lastHtmlHash) {
          console.log(`Pre-flight Cache HIT for scraper ${scraperId}. Serving cached records (0 compute cost).`);
          return {
            status: 'SUCCESS',
            records: scraper.cachedRecords as any[],
            error: null,
            cached: true,
            hash: currentHash,
          };
        }
      }
    }

    // 2. Cache Miss: Execute run
    let res: { status: string; records: any[]; error: string | null };
    if (this.shouldSimulate(collectorId)) {
      res = await this._runSimulatedCollector(collectorId, targetUrl, schema, scraperId);
    } else {
      res = await this._runRealCollector(collectorId, targetUrl);
    }

    // 3. Compute and attach hash details
    if (!currentHash) {
      currentHash = await this.computePageHash(targetUrl);
    }

    return {
      ...res,
      hash: currentHash,
      cached: false,
    };
  }

  private static async _runRealCollector(collectorId: string, targetUrl: string): Promise<{
    status: string;
    records: any[];
    error: string | null;
  }> {
    console.log(`Triggering real Bright Data collector ${collectorId} for URL ${targetUrl}`);
    const apiKey = process.env.BRIGHTDATA_API_KEY;
    
    if (apiKey) {
      try {
        const url = 'https://api.brightdata.com/dca/trigger_immediate';
        const response = await fetch(url, {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${apiKey}`,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            collector: collectorId,
            input: [{ url: targetUrl }],
          }),
        });

        if (response.status === 200) {
          let records = await response.json();
          if (records && typeof records === 'object' && 'output' in records) {
            records = records.output;
          }
          const recordsList = Array.isArray(records) ? records : [records];
          return {
            status: 'SUCCESS',
            records: recordsList,
            error: null,
          };
        } else {
          const text = await response.text();
          return {
            status: 'FAILED',
            records: [],
            error: `Bright Data API Error (Status ${response.status}): ${text}`,
          };
        }
      } catch (e: any) {
        console.error('Failed to run via API, falling back to CLI:', e);
      }
    }

    // Fallback/Alternative via CLI
    try {
      const { execSync } = require('child_process');
      const cmd = `npx -p @brightdata/cli brightdata scraper run ${collectorId} -u "${targetUrl}" --format json`;
      const stdout = execSync(cmd, { timeout: 90000 }).toString();
      const records = JSON.parse(stdout);
      const recordsList = Array.isArray(records) ? records : [records];
      
      return {
        status: 'SUCCESS',
        records: recordsList,
        error: null,
      };
    } catch (e: any) {
      return {
        status: 'FAILED',
        records: [],
        error: `CLI/System error calling collector: ${e.message}`,
      };
    }
  }

  private static async _runSimulatedCollector(
    collectorId: string,
    targetUrl: string,
    schema: Record<string, string>,
    scraperId?: string
  ): Promise<{
    status: string;
    records: any[];
    error: string | null;
  }> {
    const key = scraperId || collectorId || 'default';

    if (!SIMULATED_SELECTORS[key]) {
      // Try to load latest selectors from DB
      let dbSelectors: Record<string, string> | null = null;
      if (scraperId) {
        try {
          const latestVersion = await db.select()
            .from(selectorVersions)
            .where(eq(selectorVersions.scraperId, scraperId))
            .orderBy(desc(selectorVersions.version))
            .limit(1);
          if (latestVersion.length > 0) {
            dbSelectors = latestVersion[0].selectors as Record<string, string>;
          }
        } catch (dbErr) {
          console.error('Failed to load selectors from DB:', dbErr);
        }
      }

      if (dbSelectors) {
        SIMULATED_SELECTORS[key] = dbSelectors;
      } else {
        const template = getTemplateForUrl(targetUrl);
        const sels: Record<string, string> = {};
        for (const field of Object.keys(schema)) {
          sels[field] = template?.selectors[field] ?? `.${field}`;
        }
        SIMULATED_SELECTORS[key] = sels;
      }
    }

    const selectors = SIMULATED_SELECTORS[key];
    console.log(`Running simulated scraper run for ${key} with selectors:`, selectors);

    try {
      if (
        targetUrl.includes('localhost') ||
        targetUrl.includes('127.0.0.1') ||
        targetUrl.includes('/demo-site') ||
        targetUrl.includes('/api/demo')
      ) {
        const fetchUrl = `${this.getBaseUrl()}/api/demo/catalog-raw`;
        const res = await fetch(fetchUrl, { next: { revalidate: 0 } });
        if (res.status === 200) {
          const data = await res.json();
          const rawItems = data.items || [];
          const currentDomClasses = data.dom_classes || {};

          const parsedRecords = rawItems.map((item: any) => {
            const record: Record<string, any> = {};
            for (const field of Object.keys(schema)) {
              const selector = selectors[field] || '';
              const expectedClass = `.${currentDomClasses[field] || 'invalid-class'}`;
              
              if (selector === expectedClass) {
                const val = item[field];
                if (schema[field] === 'number' && typeof val === 'string') {
                  const cleanVal = val.replace('$', '').replace(/,/g, '').trim();
                  record[field] = parseFloat(cleanVal);
                } else {
                  record[field] = val;
                }
              } else {
                record[field] = null;
              }
            }
            return record;
          });

          return {
            status: 'SUCCESS',
            records: parsedRecords,
            error: null,
          };
        }
      }

      // Public sandbox sites (quotes.toscrape.com, books.toscrape.com)
      const sandboxRecords = await fetchSandboxRecords(targetUrl);
      if (sandboxRecords && sandboxRecords.length > 0) {
        return {
          status: 'SUCCESS',
          records: mapRecordsToSchema(sandboxRecords, schema),
          error: null,
        };
      }

      // Static fallback if not hitting local demo or known sandbox
      const staticRecords = [
        { product_name: 'Premium Keyboard', price: 129.99, currency: 'USD', availability: 'In Stock', product_url: 'https://example.com/item/1' },
        { product_name: 'Wireless Mouse', price: 49.99, currency: 'USD', availability: 'In Stock', product_url: 'https://example.com/item/2' },
        { product_name: 'USB-C Hub', price: 35.00, currency: 'USD', availability: 'Low Stock', product_url: 'https://example.com/item/3' },
      ];

      const filtered = staticRecords.map((item: any, idx: number) => {
        const record: Record<string, any> = {};
        for (const [field, expectedType] of Object.entries(schema)) {
          if (selectors[field] && !selectors[field].startsWith('.broken')) {
            let val = item[field];
            if (val === undefined) {
              if (expectedType === 'number') {
                val = 9.99 + idx * 5;
              } else if (expectedType === 'url') {
                val = `https://example.com/mock-item/${idx + 1}`;
              } else if (expectedType === 'boolean') {
                val = true;
              } else {
                val = `Mock ${field} ${idx + 1}`;
              }
            }
            record[field] = val;
          } else {
            record[field] = null;
          }
        }
        return record;
      });

      return {
        status: 'SUCCESS',
        records: filtered,
        error: null,
      };
    } catch (e: any) {
      console.error('Failed simulated collection:', e);
      return {
        status: 'FAILED',
        records: [],
        error: `Simulated collection error: ${e.message}`,
      };
    }
  }

  static async healCollector(
    collectorId: string,
    failureDescription: string,
    scraperId?: string
  ): Promise<{
    status: string;
    error: string | null;
    logs: string;
  }> {
    if (this.shouldSimulate(collectorId)) {
      return this._healSimulatedCollector(collectorId, failureDescription, scraperId);
    }
    return this._healRealCollector(collectorId, failureDescription);
  }

  private static async _healRealCollector(collectorId: string, failureDescription: string): Promise<{
    status: string;
    error: string | null;
    logs: string;
  }> {
    try {
      const { execSync } = require('child_process');
      const cmd = `npx -p @brightdata/cli brightdata scraper heal ${collectorId} --message "${failureDescription.replace(/"/g, '\\"')}"`;
      const logs = execSync(cmd, { timeout: 60000 }).toString();

      try {
        const approveCmd = `npx -p @brightdata/cli brightdata scraper approve ${collectorId}`;
        execSync(approveCmd, { timeout: 30000 });
      } catch (approveErr) {
        console.error('Approve failed:', approveErr);
      }

      return {
        status: 'SUCCESS',
        error: null,
        logs,
      };
    } catch (e: any) {
      return {
        status: 'FAILED',
        error: e.message,
        logs: '',
      };
    }
  }

  private static async _healSimulatedCollector(
    collectorId: string,
    failureDescription: string,
    scraperId?: string
  ): Promise<{
    status: string;
    error: string | null;
    logs: string;
  }> {
    const key = scraperId || collectorId || 'default';
    console.log(`Running simulated AI healing for ${key}. Diagnostics:`, failureDescription);

    try {
      let targetUrl = '';
      if (scraperId) {
        const found = await db.select().from(scrapers).where(eq(scrapers.id, scraperId)).limit(1);
        targetUrl = found[0]?.targetUrl || '';
      }

      const template = targetUrl ? getTemplateForUrl(targetUrl) : undefined;
      if (template && template.id !== 'demo-site') {
        SIMULATED_SELECTORS[key] = { ...template.selectors };
        return {
          status: 'SUCCESS',
          error: null,
          logs: `AI Healing Agent restored sandbox selectors for ${template.label}.\n` +
            Object.entries(template.selectors).map(([field, sel]) => `- ${field}: ${sel}`).join('\n'),
        };
      }

      const fetchUrl = `${this.getBaseUrl()}/api/demo/catalog-raw`;
      const res = await fetch(fetchUrl, { next: { revalidate: 0 } });
      if (res.status !== 200) {
        throw new Error(`Demo website status: ${res.status}`);
      }

      const data = await res.json();
      const currentDomClasses = data.dom_classes || {};
      const currentState = data.status || 'NORMAL';

      let healedLog = '';

      if (['DATA_DROPS', 'COUNT_COLLAPSE', 'TYPE_MISMATCH', 'EMPTY_EXTRACTION', 'QUALITY_DEGRADE'].includes(currentState)) {
        // Restore catalog structure on demo website via API call
        await fetch(`${this.getBaseUrl()}/api/demo/state`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ status: 'NORMAL' }),
        });

        SIMULATED_SELECTORS[key] = {
          product_name: '.product-title',
          price: '.product-price',
          currency: '.product-currency',
          availability: '.product-stock',
          product_url: '.product-link',
        };
        healedLog = 'AI Healing Agent resolved data structure collapse. Restored data extraction thresholds to baseline values.';
      } else {
        SIMULATED_SELECTORS[key] = {};
        for (const [field, clsName] of Object.entries(currentDomClasses)) {
          SIMULATED_SELECTORS[key][field] = `.${clsName}`;
        }
        healedLog = 'AI Healing Agent analysed HTML structure.\nSUCCESS: Outdated selectors updated.\n' +
          Object.entries(currentDomClasses).map(([field, clsName]) => `- ${field}: changed to '.${clsName}'`).join('\n');
      }

      return {
        status: 'SUCCESS',
        error: null,
        logs: healedLog,
      };
    } catch (e: any) {
      return {
        status: 'FAILED',
        error: `Simulated healing error: ${e.message}`,
        logs: '',
      };
    }
  }

  static resetSelectors(scraperId?: string, collectorId?: string, targetUrl?: string) {
    const key = scraperId || collectorId || 'default';
    const template = targetUrl ? getTemplateForUrl(targetUrl) : undefined;
    SIMULATED_SELECTORS[key] = template?.selectors ?? {
      product_name: '.product-title',
      price: '.product-price',
      currency: '.product-currency',
      availability: '.product-stock',
      product_url: '.product-link',
    };
  }

  static initSelectorsForScraper(scraperId: string, targetUrl: string) {
    this.resetSelectors(scraperId, undefined, targetUrl);
    return this.getSelectors(scraperId);
  }

  static getSelectors(scraperId?: string, collectorId?: string): Record<string, string> {
    const key = scraperId || collectorId || 'default';
    if (!SIMULATED_SELECTORS[key]) {
      this.resetSelectors(scraperId, collectorId);
    }
    return SIMULATED_SELECTORS[key];
  }

  static setSelectors(selectors: Record<string, string>, scraperId?: string, collectorId?: string) {
    const key = scraperId || collectorId || 'default';
    SIMULATED_SELECTORS[key] = selectors;
  }
}

File: apps/api/services/brightdata.py
import os
import subprocess
import json
import httpx
import logging
from typing import Dict, Any, List, Optional
from urllib.parse import urlparse

logger = logging.getLogger("scrapeguard.brightdata")

# Simulated selector store to demonstrate selector-based healing locally
# Format: { scraper_id_or_collector_id: { field_name: selector_definition } }
SIMULATED_SELECTORS = {}

class BrightDataService:
    @classmethod
    def is_mock_mode(cls) -> bool:
        # Default to mock/simulation mode if Bright Data credentials are not set
        api_key = os.getenv("BRIGHTDATA_API_KEY")
        # Or if explicitly requested in environment variables
        mock_env = os.getenv("MOCK_BRIGHTDATA", "true").lower() == "true"
        return not api_key or mock_env

    @classmethod
    def compute_page_hash(cls, url: str) -> Optional[str]:
        """
        Computes a SHA256 hash of the target webpage content to determine structural layout changes.
        """
        try:
            import hashlib
            if cls.is_mock_mode():
                # In simulation mode, fetch the local catalog mock raw endpoint
                local_url = "http://127.0.0.1:8000/api/demo/catalog-raw"
                try:
                    response = httpx.get(local_url, timeout=5.0)
                    if response.status_code == 200:
                        return hashlib.sha256(response.content).hexdigest()
                except Exception:
                    # Local server is offline (e.g. during test run), fallback to stable hash of the URL
                    return hashlib.sha256(url.encode('utf-8')).hexdigest()
            else:
                response = httpx.get(url, timeout=10.0, follow_redirects=True)
                if response.status_code == 200:
                    return hashlib.sha256(response.content).hexdigest()
        except Exception as e:
            logger.error(f"Error computing page hash: {str(e)}")
        return None

    @classmethod
    def update_scraper_cache(cls, db, scraper_id: str, records: List[Dict[str, Any]], html_hash: str):
        """
        Helper to save successfully validated extraction results and their HTML hash.
        """
        if not html_hash or not db:
            return
        from database.models import Scraper
        scraper = db.query(Scraper).filter(Scraper.id == scraper_id).first()
        if scraper:
            scraper.last_html_hash = html_hash
            scraper.cached_records = records
            db.commit()

    @classmethod
    def run_collector(
        cls,
        collector_id: str,
        target_url: str,
        schema: Dict[str, str],
        scraper_id: str = None,
        db = None,
        bypass_cache: bool = False
    ) -> Dict[str, Any]:
        """
        Executes a scraper collector against a target URL, performing a pre-flight cache hit check first.
        """
        current_hash = None
        
        # 1. Pre-flight Cache Check
        if db and scraper_id and not bypass_cache:
            from database.models import Scraper
            scraper = db.query(Scraper).filter(Scraper.id == scraper_id).first()
            if scraper and scraper.last_html_hash and scraper.cached_records:
                current_hash = cls.compute_page_hash(target_url)
                if current_hash and current_hash == scraper.last_html_hash:
                    logger.info(f"Pre-flight Cache HIT for scraper {scraper_id}. Serving cached records (0 compute cost).")
                    return {
                        "status": "SUCCESS",
                        "records": scraper.cached_records,
                        "error": None,
                        "cached": True,
                        "hash": current_hash
                    }

        # 2. Cache Miss: Execute run
        if cls.is_mock_mode():
            res = cls._run_simulated_collector(collector_id, target_url, schema, scraper_id, db=db)
        else:
            res = cls._run_real_collector(collector_id, target_url)

        # 3. Compute and attach hash details
        if not current_hash:
            current_hash = cls.compute_page_hash(target_url)
            
        res["hash"] = current_hash
        res["cached"] = False
        return res

    @classmethod
    def _run_real_collector(cls, collector_id: str, target_url: str) -> Dict[str, Any]:
        """
        Runs a real Bright Data collector using their CLI or Collection API.
        """
        logger.info(f"Triggering real Bright Data collector {collector_id} for URL {target_url}")
        
        # 1. Option A: Try running via the Collection API (Trigger Immediate)
        api_key = os.getenv("BRIGHTDATA_API_KEY")
        
        if api_key:
            try:
                # Bright Data DCA Endpoint: POST https://api.brightdata.com/dca/trigger_immediate
                url = "https://api.brightdata.com/dca/trigger_immediate"
                headers = {
                    "Authorization": f"Bearer {api_key}",
                    "Content-Type": "application/json"
                }
                payload = {
                    "collector": collector_id,
                    "input": [{"url": target_url}]
                }
                
                response = httpx.post(url, headers=headers, json=payload, timeout=60.0)
                if response.status_code == 200:
                    records = response.json()
                    # Some responses are wrapped inside output containers
                    if isinstance(records, dict) and "output" in records:
                        records = records["output"]
                    if not isinstance(records, list):
                        records = [records]
                    
                    return {
                        "status": "SUCCESS",
                        "records": records,
                        "error": None
                    }
                else:
                    return {
                        "status": "FAILED",
                        "records": [],
                        "error": f"Bright Data API Error (Status {response.status_code}): {response.text}"
                    }
            except Exception as e:
                logger.error(f"Failed to run via API, falling back to CLI: {e}")

        # 2. Option B: Fallback/Alternative via CLI
        try:
            # Command: bdata scraper run <collector_id> -u <url>
            # We fetch raw stdout as JSON
            cmd = ["npx", "-p", "@brightdata/cli", "brightdata", "scraper", "run", collector_id, "-u", target_url, "--format", "json"]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=90)
            
            if result.returncode == 0:
                try:
                    records = json.loads(result.stdout)
                    if not isinstance(records, list):
                        records = [records]
                    return {
                        "status": "SUCCESS",
                        "records": records,
                        "error": None
                    }
                except json.JSONDecodeError:
                    return {
                        "status": "FAILED",
                        "records": [],
                        "error": f"Failed to parse JSON from CLI stdout: {result.stdout[:200]}"
                    }
            else:
                return {
                    "status": "FAILED",
                    "records": [],
                    "error": f"CLI error: {result.stderr or result.stdout}"
                }
        except Exception as e:
            return {
                "status": "FAILED",
                "records": [],
                "error": f"System error calling CLI: {str(e)}"
            }

    @classmethod
    def heal_collector(
        cls,
        collector_id: str,
        failure_description: str,
        scraper_id: str = None
    ) -> Dict[str, Any]:
        """
        Triggers AI self-healing on a collector.
        """
        if cls.is_mock_mode():
            return cls._heal_simulated_collector(collector_id, failure_description, scraper_id)
        
        return cls._heal_real_collector(collector_id, failure_description)

    @classmethod
    def _heal_real_collector(cls, collector_id: str, failure_description: str) -> Dict[str, Any]:
        """
        Calls Bright Data CLI command to heal the collector.
        """
        try:
            logger.info(f"Triggering real Bright Data CLI heal for collector {collector_id}")
            cmd = ["npx", "-p", "@brightdata/cli", "brightdata", "scraper", "heal", collector_id, "--message", failure_description]
            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
            
            if result.returncode == 0:
                # Approve the heals automatically
                approve_cmd = ["npx", "-p", "@brightdata/cli", "brightdata", "scraper", "approve", collector_id]
                subprocess.run(approve_cmd, capture_output=True, text=True, timeout=30)
                
                return {
                    "status": "SUCCESS",
                    "error": None,
                    "logs": result.stdout
                }
            else:
                return {
                    "status": "FAILED",
                    "error": result.stderr or result.stdout,
                    "logs": result.stdout
                }
        except Exception as e:
            return {
                "status": "FAILED",
                "error": f"System error calling heal CLI: {str(e)}",
                "logs": ""
            }

    # --- SIMULATOR MODE LOGIC ---
    
    @classmethod
    def _run_simulated_collector(
        cls,
        collector_id: str,
        target_url: str,
        schema: Dict[str, str],
        scraper_id: str = None,
        db = None
    ) -> Dict[str, Any]:
        """
        Simulated scraper run. Fetches the page content (from our local mock catalog)
        and parses it using selectors stored in SIMULATED_SELECTORS.
        """
        key = scraper_id or collector_id or "default"
        
        # Initialize default selectors if not set
        if key not in SIMULATED_SELECTORS:
            # Try to load latest selectors from database
            if db and scraper_id:
                try:
                    from database.models import SelectorVersion
                    latest_sv = db.query(SelectorVersion).filter(
                        SelectorVersion.scraper_id == scraper_id
                    ).order_by(SelectorVersion.version.desc()).first()
                    if latest_sv:
                        SIMULATED_SELECTORS[key] = latest_sv.selectors
                except Exception as e:
                    logger.error(f"Failed to load selectors from DB: {e}")

            if key not in SIMULATED_SELECTORS:
                sels = {}
                for field in schema:
                    if field == "product_name": sels[field] = ".product-title"
                    elif field == "price": sels[field] = ".product-price"
                    elif field == "currency": sels[field] = ".product-currency"
                    elif field == "availability": sels[field] = ".product-stock"
                    elif field == "product_url": sels[field] = ".product-link"
                    else: sels[field] = f".{field}"
                SIMULATED_SELECTORS[key] = sels

        selectors = SIMULATED_SELECTORS[key]
        logger.info(f"Running simulated scraper run for {key} with selectors: {selectors}")

        try:
            # Check target url. If it is pointing to our mock catalog API, fetch it
            # Otherwise we mock a static response
            if "localhost" in target_url or "127.0.0.1" in target_url or "/demo-site" in target_url or "/api/demo" in target_url:
                try:
                    # Let's request the endpoint from our local server
                    # If target url points to html demo site, redirect to raw JSON catalog data
                    request_url = target_url
                    if "/demo-site" in request_url:
                        request_url = request_url.replace("/demo-site", "/api/demo/catalog-raw")
                    response = httpx.get(request_url, timeout=10.0)
                    if response.status_code == 200:
                        # The mock demo site returns structured mock HTML,
                        # but it also returns a JSON config of its *current* structure state so we can simulate CSS renaming!
                        data = response.json()
                        raw_items = data.get("items", [])
                        current_dom_classes = data.get("dom_classes", {})
                        
                        # Apply parsing logic based on selectors compared to the current DOM classes
                        parsed_records = []
                        for item in raw_items:
                            record = {}
                            for field in schema:
                                # Map schema field to CSS selector
                                selector = selectors.get(field, "")
                                expected_class = f".{current_dom_classes.get(field, 'invalid-class')}"
                                if selector == expected_class:
                                    # Success match
                                    val = item.get(field)
                                    # Simulate type casting based on schema
                                    if schema[field] == "number" and isinstance(val, str):
                                        try:
                                            # Clean currency symbol if any
                                            clean_val = val.replace("$", "").replace(",", "").strip()
                                            record[field] = float(clean_val)
                                        except Exception:
                                            record[field] = val
                                    else:
                                        record[field] = val
                                else:
                                    # Failure: Website class changed, selector broke!
                                    record[field] = None
                                    
                            parsed_records.append(record)
                            
                        return {
                            "status": "SUCCESS",
                            "records": parsed_records,
                            "error": None
                        }
                except Exception as e:
                    logger.error(f"Failed to query local mock catalog: {e}")
            
            # Static mock fallback if not running local demo site URL
            static_records = [
                {"product_name": "Premium Keyboard", "price": 129.99, "currency": "USD", "availability": "In Stock", "product_url": "https://example.com/item/1"},
                {"product_name": "Wireless Mouse", "price": 49.99, "currency": "USD", "availability": "In Stock", "product_url": "https://example.com/item/2"},
                {"product_name": "USB-C Hub", "price": 35.00, "currency": "USD", "availability": "Low Stock", "product_url": "https://example.com/item/3"},
            ]
            
            # Filter through simulated selectors
            filtered = []
            for idx, item in enumerate(static_records):
                record = {}
                for field in schema:
                    # If selector is present, assume it extracts it. Otherwise it is null
                    if field in selectors and not selectors[field].startswith(".broken"):
                        val = item.get(field)
                        if val is None:
                            expected_type = schema[field]
                            if expected_type == "number":
                                val = 9.99 + idx * 5.0
                            elif expected_type == "url":
                                val = f"https://example.com/mock-item/{idx + 1}"
                            elif expected_type == "boolean":
                                val = True
                            else:
                                val = f"Mock {field} {idx + 1}"
                        record[field] = val
                    else:
                        record[field] = None
                filtered.append(record)
                
            return {
                "status": "SUCCESS",
                "records": filtered,
                "error": None
            }
        except Exception as e:
            return {
                "status": "FAILED",
                "records": [],
                "error": f"Simulation failure: {str(e)}"
            }

    @classmethod
    def _heal_simulated_collector(
        cls,
        collector_id: str,
        failure_description: str,
        scraper_id: str = None
    ) -> Dict[str, Any]:
        """
        Simulated Healing:
        Analyzes the failure description and updates the SIMULATED_SELECTORS to match the website's updated CSS classes.
        In a real scenario, Bright Data AI self-heals it. In our simulation, we query the demo site's active DOM classes
        and "heal" our selectors to match them.
        """
        key = scraper_id or collector_id or "default"
        logger.info(f"Running simulated AI healing for {key}. Diagnostics:\n{failure_description}")

        try:
            from routes.demo import DEMO_WEBSITE_STATE
            current_state = DEMO_WEBSITE_STATE["status"]
            
            # If data is dropped/collapsed, simulate structural repair by restoring healthy catalog state
            if current_state in ["DATA_DROPS", "COUNT_COLLAPSE", "TYPE_MISMATCH", "EMPTY_EXTRACTION", "QUALITY_DEGRADE"]:
                DEMO_WEBSITE_STATE["status"] = "NORMAL"
                DEMO_WEBSITE_STATE["dom_classes"] = {
                    "product_name": "product-title",
                    "price": "product-price",
                    "currency": "product-currency",
                    "availability": "product-stock",
                    "product_url": "product-link"
                }
                SIMULATED_SELECTORS[key] = {
                    "product_name": ".product-title",
                    "price": ".product-price",
                    "currency": ".product-currency",
                    "availability": ".product-stock",
                    "product_url": ".product-link"
                }
                healed_log = f"AI Healing Agent resolved data structure collapse. Restored data extraction thresholds to baseline values."
            else:
                # Class renaming or element moved: dynamically map current classes to selectors
                current_classes = DEMO_WEBSITE_STATE["dom_classes"]
                SIMULATED_SELECTORS[key] = {
                    field: f".{cls_name}"
                    for field, cls_name in current_classes.items()
                }
                healed_log = "AI Healing Agent analysed HTML structure.\nSUCCESS: Outdated selectors updated.\n" + "\n".join([
                    f"- {field}: changed to '.{cls_name}'" for field, cls_name in current_classes.items()
                ])

            return {
                "status": "SUCCESS",
                "error": None,
                "logs": healed_log
            }
        except Exception as e:
            return {
                "status": "FAILED",
                "error": f"Simulated healing error: {str(e)}",
                "logs": ""
            }

    @classmethod
    def reset_selectors(cls, scraper_id: str = None, collector_id: str = None):
        """
        Helper to break the selectors back to default, useful for resetting the demo.
        """
        key = scraper_id or collector_id or "default"
        SIMULATED_SELECTORS[key] = {
            "product_name": ".product-title",
            "price": ".product-price",
            "currency": ".product-currency",
            "availability": ".product-stock",
            "product_url": ".product-link"
        }

    @classmethod
    def get_selectors(cls, scraper_id: str = None, collector_id: str = None) -> Dict[str, str]:
        key = scraper_id or collector_id or "default"
        if key not in SIMULATED_SELECTORS:
            cls.reset_selectors(scraper_id=scraper_id, collector_id=collector_id)
        return SIMULATED_SELECTORS[key]

    @classmethod
    def set_selectors(cls, selectors: Dict[str, str], scraper_id: str = None, collector_id: str = None):
        key = scraper_id or collector_id or "default"
        SIMULATED_SELECTORS[key] = selectors
